### PR TITLE
Add log_likelihood to mcmc and vi samples

### LINKED ISSUE #292: Add log likelihood to arviz trace objects
I tried to run the `loo` function from arviz and got an exception because that part of the model was not annotated in the trace:

```
data_loo = az.loo(tr, pointwise=True)

TypeError: log likelihood not found in inference data object

```

### PR DESCRIPTION
**Pull request by Sayam753**
Resolves #292 

Please review this PR, then I can write tests.

### CODE REVIEW

File: pymc4/mcmc/utils.py [hunk 1]
```diff
@@ -127,12 +129,17 @@ def trace_to_arviz(
...
+        }
 
     return az.from_dict(
         posterior=trace,
         sample_stats=sample_stats,
         prior_predictive=prior_predictive,
         posterior_predictive=posterior_predictive,
+        log_likelihood=log_likelihood,
         observed_data=observed_data,
     )
```
Comment (Sayam753) [inline pymc4/mcmc/utils.py:144]: I am a bit confused here whether log_likelihood has to be a group or should it be a variable in `sample_stats`.

Comment (Sayam753) [reply pymc4/mcmc/utils.py:144]: As per discussions in ArviZ gitter channel, it's better to have `log_likelihood` as a group for compatibility with Multi-observed random variables. So, I will not change `trace_to_arviz` function.

**Review by junpenglao — APPROVED** [review]

**Comment by Sayam753** [comment]
Thanks @junpenglao . I am writing a few tests for this and hopefully push here soon.

File: pymc4/mcmc/samplers.py [hunk 2]
```diff
@@ -822,3 +827,28 @@ def vectorized_logpfn(*state):
...
+
+    def extract_log_likelihood(values, observed_rv):
+        st = flow.SamplingState.from_values(values, observed_values=sampling_state.observed_values)
+        _, st = flow.evaluate_model_transformed(model, state=st)
+        try:
+            dist = st.continuous_distributions[observed_rv]
+        except KeyError:
+            dist = st.discrete_distributions[observed_rv]
+        return dist.log_prob(dist.model_info["observed"])
+
```
Comment (Sayam753) [inline pymc4/mcmc/samplers.py:845]: I am not sure here about having deterministic distributions as observed variables. If that's the case, we do need to handle that as well.

**Review by Sayam753 — COMMENTED** [review]
This PR is all good to go from my side.

**Comment by Sayam753** [comment]
This is interesting. Latest stable release of `nbconvert>=6.0.0` causes the notebooks tests to fail.

**Comment by mattiasthalen** [comment]
When running pm.sample_posterior_predictive, that idata group is not generated. Is that covered by this fix?

**Comment by Sayam753** [comment]
Hi @mattiasthalen 
Sorry for getting back late on this. I checked with both mcmc and vi, that after passing trace (which includes log likelihood) to posterior predictive, the log_likelihood group is still there. 
Can you share a reproducible code snippet, if the issue still persists.
Thanks

**Comment by mattiasthalen** [comment]
> Hi @mattiasthalen
> Sorry for getting back late on this. I checked with both mcmc and vi, that after passing trace (which includes log likelihood) to posterior predictive, the log_likelihood group is still there.
> Can you share a reproducible code snippet, if the issue still persists.
> Thanks

No worries! The mistake was all mine, I forgot to preface the likelihood function with yield... doh! ☺️

**Pull request merged**